Add a flag to get amount results as integer

oxarbitrage opened the pull request:
Closes https://github.com/zcash/zcash/issues/2375

oxarbitrage commented:
Bellow is a list of modified rpc methods where amount is returned inside `rpc_wallet.cpp`. 
4 methods need an optional boolean argument `inZat` to get the amount in zatoshis, this is needed as these functions currently only return a number. In the rest `amountZat` field was added everywhere after `amount` output.

- `getreceivedbyaddress` - bool arg `inZat` added.
- `getreceivedbyaccount` - bool arg `inZat` added.
- `getbalance` - bool arg `inZat` added.
- `listreceivedbyaddress` - output field `amountZat` added.
- `listreceivedbyaccount` - output field `amountZat` added.
- `listtransactions` - output field `amountZat` added.
- `listsinceblock` - output field `amountZat` added.
- `gettransaction` - output field `amountZat` added.
- `listunspent` - output field `amountZat` added.
- `z_listreceivedbyaddress` - output field `amountZat` added.
- `z_getbalance` - bool arg `inZat` added.

daira commented on qa/rpc-tests/wallet.py:
```diff
@@ -381,15 +388,18 @@ def run_test (self):
         txId  = self.nodes[0].sendtoaddress(self.nodes[2].getnewaddress(), "1e-4")
         txObj = self.nodes[0].gettransaction(txId)
         assert_equal(txObj['amount'], Decimal('-0.00010000'))
+        assert_equal(txObj['amountZat'], int('-10000'))
```
```suggestion
        assert_equal(txObj['amountZat'], -10000)
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -633,7 +634,12 @@ UniValue getreceivedbyaddress(const UniValue& params, bool fHelp)
 
-    return  ValueFromAmount(nAmount);
+    // inZat
+    if (params.size() > 2)
+        if (params[2].get_bool())
+            return nAmount;
```
```suggestion
    if (params.size() > 2 && params[2].get_bool()) {
        return nAmount;
    }
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -690,6 +697,11 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
     }
 
+    // inZat
+    if (params.size() > 2)
+        if (params[2].get_bool())
+            return nAmount;
```
```suggestion
    if (params.size() > 2 && params[2].get_bool()) {
        return nAmount;
    }
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -788,6 +801,12 @@ UniValue getbalance(const UniValue& params, bool fHelp)
         }
+
+        // inZat
+        if (params.size() > 3)
+            if (params[3].get_bool())
+                return nBalance;
```
```suggestion
        if (params.size() > 3 && params[3].get_bool()) {
            return nAmount;
        }
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1444,6 +1470,7 @@ UniValue listtransactions(const UniValue& params, bool fHelp)
             "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and for the\n"
             "                                         'move' category for moves outbound. It is positive for the 'receive' category,\n"
             "                                         and for the 'move' category for inbound funds.\n"
+            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as 'amount' field.\n"
```
```suggestion
            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as for the 'amount' field.\n"
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1639,6 +1666,7 @@ UniValue listsinceblock(const UniValue& params, bool fHelp)
             "    \"category\":\"send|receive\",     (string) The transaction category. 'send' has negative amounts, 'receive' has positive amounts.\n"
             "    \"amount\": x.xxx,          (numeric) The amount in " + CURRENCY_UNIT + ". This is negative for the 'send' category, and for the 'move' category for moves \n"
             "                                          outbound. It is positive for the 'receive' category, and for the 'move' category for inbound funds.\n"
+            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as 'amount' field.\n"
```
```suggestion
            "    \"amountZat\": x.xxx,       (numeric) The amount in zatoshis. Negative and positive are the same as for the 'amount' field.\n"
```

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -1268,6 +1289,7 @@ UniValue listreceivedbyaddress(const UniValue& params, bool fHelp)
             "    \"address\" : \"receivingaddress\",  (string) The receiving address\n"
             "    \"account\" : \"accountname\",       (string) DEPRECATED. The account of the receiving address. The default account is \"\".\n"
             "    \"amount\" : x.xxx,                  (numeric) The total amount in " + CURRENCY_UNIT + " received by the address\n"
+            "    \"amountZat\" : xxxx                 (numeric) The amount in zatoshis\n"
```
Add `MINOR_CURRENCY_UNIT` set to "zatoshis", analogous to `CURRENCY_UNIT`. (The "Zat" suffixes on field names don't need to change.) This could be a separate PR if you prefer.

daira commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3461,6 +3498,11 @@ UniValue z_getbalance(const UniValue& params, bool fHelp)
     }
 
+    // inZat
+    if (params.size() > 3)
+        if (params[3].get_bool())
+            return nBalance;
```
```suggestion
    if (params.size() > 3 && params[3].get_bool()) {
        return nBalance;
    }
```

oxarbitrage commented:
Please don't merge, working in the suggested changes.

oxarbitrage commented:
the comments and suggestions are all done in the last 5 commits.

daira commented on qa/rpc-tests/receivedby.py:
```diff
@@ -73,19 +74,19 @@ def run_test(self):
         check_array_result(self.nodes[1].listreceivedbyaddress(),
                            {"address":addr},
-                           {"address":addr, "account":"", "amount":Decimal("0.1"), "confirmations":10, "txids":[txid,]})
+                           {"address":addr, "account":"", "amount":Decimal("0.1"), "amountZat":int("10000000"), "confirmations":10, "txids":[txid,]})
```
The integers in this file don't need `int()`. (Does not block.)

daira commented on qa/rpc-tests/wallet.py:
```diff
@@ -222,6 +222,7 @@ def run_test (self):
             if uTx['txid'] == zeroValueTxid:
                 found = True
                 assert_equal(uTx['amount'], Decimal('0.00000000'))
+                assert_equal(uTx['amountZat'], int('0'))
```
Same here, no need for `int()`.

daira commented on qa/rpc-tests/wallet_listreceived.py:
```diff
@@ -16,6 +16,7 @@
 no_memo = 'f6' + ('0'*1022) # see section 5.5 of the protocol spec
 
 fee = Decimal('0.0001')
+feeZat = Decimal('10000')
```
No need for `Decimal` here, just use integers.

daira commented on qa/rpc-tests/wallet_listreceived.py:
```diff
@@ -71,11 +73,13 @@ def run_test_release(self, release, height):
 
         assert_equal(txid, r[0]['txid'])
         assert_equal(Decimal('0.4')-fee, r[0]['amount'])
+        assert_equal(Decimal('40000000')-feeZat, r[0]['amountZat'])
```
and here.

oxarbitrage commented:
I rebased and added a new commit to fix the last comments. Let me know how do you see it. Thank you @daira

daira commented on qa/rpc-tests/wallet.py:
```diff
@@ -263,10 +264,13 @@ def run_test (self):
         assert_equal(self.nodes[2].getbalance("*"), Decimal('13.99800000')) #should not be
 
+        # check integer balances from getbalance
+        assert_equal(self.nodes[2].getbalance("*", 1, False, True), 1399800000) #should not be
```
Nonblocking: does anyone know what "should not be?" means in these three comments?

daira replied:
Seems like it has been there since the code was added in https://github.com/zcash/zcash/commit/77650cc9f4646787e843b2922730d8cba5af36a1 . I suggest just removing it (on 3 lines).

oxarbitrage replied:
Removed comments at https://github.com/zcash/zcash/pull/4282/commits/76cb4e8d0437be9f2843d7affac0f7573605dd12

str4d commented on qa/rpc-tests/receivedby.py:
```diff
@@ -71,19 +72,19 @@ def run_test(self):
         check_array_result(self.nodes[1].listreceivedbyaddress(0,True),
                            {"address":addr},
-                           {"address":addr, "account":"", "amount":0, "confirmations":0, "txids":[]})
+                           {"address":addr, "account":"", "amount":0, "confirmations":0, "amountZat":int("0"), "txids":[]})
```
```suggestion
                           {"address":addr, "account":"", "amount":0, "confirmations":0, "amountZat":0, "txids":[]})
```

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -597,13 +597,14 @@ UniValue getreceivedbyaddress(const UniValue& params, bool fHelp)
             "1. \"zcashaddress\"  (string, required) The Zcash address for transactions.\n"
-            "2. minconf             (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "2. minconf         (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
```
```suggestion
            "3. inZat           (bool, optional, default=false) Get the result amount in " + MINOR_CURRENCY_UNIT + " (as an integer).\n"
```

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -657,13 +663,14 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
             "1. \"account\"      (string, required) MUST be set to the empty string \"\" to represent the default account. Passing any other string will result in an error.\n"
-            "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "2. minconf        (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
```
Ditto.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -746,14 +758,15 @@ UniValue getbalance(const UniValue& params, bool fHelp)
             "1. \"account\"      (string, optional) DEPRECATED. If provided, it MUST be set to the empty string \"\" or to the string \"*\", either of which will give the total available balance. Passing any other string will result in an error.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
             "3. includeWatchonly (bool, optional, default=false) Also include balance in watchonly addresses (see 'importaddress')\n"
+            "4. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
```
Ditto.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3451,15 +3487,16 @@ UniValue z_getbalance(const UniValue& params, bool fHelp)
             "\nArguments:\n"
             "1. \"address\"      (string) The selected address. It may be a transparent or private address.\n"
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
```
Ditto.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -597,13 +597,14 @@ UniValue getreceivedbyaddress(const UniValue& params, bool fHelp)
+            "2. minconf         (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount   (numeric) The total amount in " + CURRENCY_UNIT + " received at this address.\n"
```
This needs to be updated to reflect the units variability.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -657,13 +663,14 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
+            "2. minconf        (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat           (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
```
Ditto.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -746,14 +758,15 @@ UniValue getbalance(const UniValue& params, bool fHelp)
             "3. includeWatchonly (bool, optional, default=false) Also include balance in watchonly addresses (see 'importaddress')\n"
+            "4. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
```
Ditto.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3451,15 +3487,16 @@ UniValue z_getbalance(const UniValue& params, bool fHelp)
             "2. minconf          (numeric, optional, default=1) Only include transactions confirmed at least this many times.\n"
+            "3. inZat            (bool, optional, default=false) Get the result amount as an integer.\n"
             "\nResult:\n"
             "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this address.\n"
```
Ditto.

oxarbitrage commented:
Add the parameters to `src/rpc/client.cpp` is needed otherwise will not work for the cli. Thank you @str4d . I added them at https://github.com/zcash/zcash/pull/4282/commits/93afebeef43597610ef8bad66c1ae1c9e69926b7 

While testing that i found some errors that were also fixed in the same commit.

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -670,9 +670,9 @@ UniValue getreceivedbyaccount(const UniValue& params, bool fHelp)
+            "3. inZat           (bool, optional, default=false) Get the result amount in " + MINOR_CURRENCY_UNIT + " (as an integer).\n"
             "\nResult:\n"
-            "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
+            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received at this address.\n"
```
Copy-paste error:
```suggestion
            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received for this account.\n"
```

str4d commented on src/wallet/rpcwallet.cpp:
```diff
@@ -766,9 +766,9 @@ UniValue getbalance(const UniValue& params, bool fHelp)
+            "4. inZat            (bool, optional, default=false) Get the result amount in " + MINOR_CURRENCY_UNIT + " (as an integer).\n"
             "\nResult:\n"
-            "amount              (numeric) The total amount in " + CURRENCY_UNIT + " received for this account.\n"
+            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received at this address.\n"
```
Copy-paste error, but here we can improve the message (since accounts are deprecated):
```suggestion
            "amount              (numeric) The total amount in " + CURRENCY_UNIT + "(or " + MINOR_CURRENCY_UNIT + " if inZat is true) received.\n"
```

oxarbitrage commented:
Thank you @str4d . I used the github "add to batch" feature to include your suggestions. Hope is ok.

str4d commented:
@zkbot r+

rex4539 commented on qa/rpc-tests/listtransactions.py:
```diff
@@ -6,6 +6,7 @@
 # Exercise the listtransactions API
 
 from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
```
```suggestion
```

daira commented:
utACK. @zkbot r+
